projects/bp-gallery/src/components/views/picture/sidebar/picture-info/TagSelectionField.tsx

```diff
@@ -65,16 +111,209 @@ const TagSelectionField = <T extends TagFields>({
+    ) {
+      // if yes then just add it to the selected tags
+      const existingTag = tagChildTags[lastAddedTag!.id].find(
+        tag => tag.name === addTag.name
+      ) as unknown as T;
+      const filteredNewValues = newValue.filter(val => !val.createValue);
+      filteredNewValues.push(existingTag);
+      newValue = filteredNewValues;
+    } else {
+      // if not create a new tag in the corresponding position
+      const { data } = await createChildMutation!({
+        variables: { name: addTag.createValue, parentIDs: [lastAddedTag!.id] },
+      });
+      handleAfterCreate(addTag, data);
+    }
+  };
+
+  const addNewSiblingTag = async (addTag: T, newValue: T[]) => {
+    if (
+      tagSiblingTags &&
+      lastAddedTag!.id in tagSiblingTags &&
+      tagSiblingTags[lastAddedTag!.id].some(tag => tag.name === addTag.name)
+    ) {
+      // if yes just add it
+      const existingTag = tagSiblingTags[lastAddedTag!.id].find(
+        tag => tag.name === addTag.name
+      ) as unknown as T;
+      const filteredNewValues = newValue.filter(val => !val.createValue);
+      filteredNewValues.push(existingTag);
+      newValue = filteredNewValues;
+    } else if (lastAddedTag!.name !== addTag.name) {
+      // if not create a new tag
+      const { data } = await createChildMutation!({
+        variables: {
+          name: addTag.createValue,
+          parentIDs:
+            tagSupertagList && lastAddedTag!.id in tagSupertagList
+              ? tagSupertagList[lastAddedTag!.id]
+                  .filter(path => path.length)
+                  .map(path => path[path.length - 1].id)
+              : [],
+          root:
+            tagSupertagList &&
+            lastAddedTag!.id in tagSupertagList &&
+            tagSupertagList[lastAddedTag!.id].some(path => !path.length),
+        },
+      });
+      handleAfterCreate(addTag, data, false, true);
+    } else {
+      const filteredNewValues = newValue.filter(val => !val.createValue);
+      newValue = filteredNewValues;
+    }
+  };
+
+  const addNewRootTag = async (addTag: T, newValue: T[]) => {
+    if (tagTree?.some(rootTag => rootTag.name === addTag.name)) {
+      const existingTag = tagTree.find(tag => tag.name === addTag.name) as unknown as T;
+      const filteredNewValues = newValue.filter(val => !val.createValue);
+      filteredNewValues.push(existingTag);
+      newValue = filteredNewValues;
+    } else {
+      const { data } = await createMutation!({
+        variables: { name: addTag.createValue },
+      });
+      handleAfterCreate(addTag, data, true);
+    }
+  };
+
   if (role >= AuthRole.CURATOR) {
     return (
       <div className='tag-selection'>
         <Autocomplete<T, true>
           multiple
           autoHighlight
-          isOptionEqualToValue={(option, value) => option.name === value.name}
-          options={tagList}
+          isOptionEqualToValue={(option, value) => option.id === value.id}
+          options={options}
```

Comment:
For me, adding a tag crashes due to [this code](https://github.com/mui/material-ui/blob/90f2f7962537ce2e8cced871f89d9255dadcb5c4/packages/mui-base/src/useAutocomplete/useAutocomplete.js#L267-L268): The `Autocomplete` detects that the new tag isn't part of `options` (which is true for a short moment, I believe) and wants to `warn` in the console. To further describe the problem, the code calls `JSON.stringify` on the tag that is in `value` (`tags`) but not in `options`. This fails, because the tags have references up and down the graph, causing `JSON.stringify` to error with `TypeError: cyclic object value`. A potential solution may be passing `unionWith(options, tags, (a, b) => a.id === b.id)` as `options` (maybe use `useMemo` for it).